Hive: Allow auto conversion of Hive types when the CREATE TABLE statement contains a not supported type

pvary opened the pull request:
To allow existing Hive workloads to easily move to Iceberg tables we should make it easier to migrate the scripts.
Currently the following types are not allowed in Hive `CREATE TABLE` statements since they do not have an exact match in Iceberg types:
- TINYINT
- SMALLINT
- VARCHAR
- CHAR

We still want to keep the 1-on-1 schema mapping between Hive and Iceberg, but for the table creation we can convert unsupported types to supported ones:
- TINYINT - INTEGER
- SMALLINT - INTERER
- VARCHAR - STRING
- CHAR - STRING

So this PR introduces a new configuration key: `iceberg.mr.schema.auto.conversion` which if it set to `true` then the `CREATE TABLE` query will not throw an exception, but will convert the columns instead.
For example:
```
CREATE EXTERNAL TABLE not_supported_types (not_supported TINYINT);

DESCRIBE not_supported_types;
----------------
not_supported | int | from deserializer
----------------
```

pvary commented:
@shardulm94: Could you spare some time to review this change?

marton-bod commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java:
```diff
@@ -73,9 +75,13 @@ Type convertType(TypeInfo typeInfo) {
-                ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory() +
-                ") for Iceberg tables. Consider using INT/INTEGER type instead.");
+            if (autoConvert) {
+              return Types.IntegerType.get();
```
can we add some logging so we can track where autoconversion happened? should be useful for debugging

pvary replied:
Good idea.
Added the comment

marton-bod reviewed (commented):
LGTM, two small comments

marton-bod commented:
Once done, we should also update the Hive docs with this new property/behaviour.

pvary commented on mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:
```diff
@@ -491,5 +497,17 @@ public void testCreateTableWithNotSupportedTypes() {
       );
     }
+
+    shell.executeStatement("SET " + InputFormatConfig.SCHEMA_AUTO_CONVERSION + "=true");
```
Definitely!
Fixed

marton-bod reviewed (approved):
LGTM, thanks for the fixes @pvary

rdblue commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaConverter.java:
```diff
@@ -73,9 +79,14 @@ Type convertType(TypeInfo typeInfo) {
+              throw new IllegalArgumentException("Unsupported Hive type (" +
+                  ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory() +
+                  ") for Iceberg tables. Consider using INT/INTEGER type instead.");
+            }
```
You could probably make this a bit cleaner by using a `Precondition`:

```java
    Preconditions.check(autoConvert,
        "Unsupported Hive type: %s, use integer instead", ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory());
    return Types.IntegerType.get();
```

That way, you'd get string formatting for free.

pvary replied:
The code is definitely sorter, if a little bit strange for me 😄 
Changed anyway

rdblue commented on mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergSerDe.java:
```diff
@@ -129,10 +130,12 @@ public ObjectInspector getObjectInspector() {
    * it adds the previousException as a root cause.
    * @param serDeProperties The source of the hive schema
    * @param previousException If we had an exception previously
+   * @param autoConversion If <code>true</code> then autoConversion of the schema is enabled
```
Since the argument name is `autoConversion`, the doc doesn't add much value. What about clarifying what it means? Like "When true, convert unsupported types to more permissive ones, like tinyint to int"

pvary replied:
Thanks for the rewording! Your one is definitely better.

rdblue commented on mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:
```diff
@@ -491,5 +497,17 @@ public void testCreateTableWithNotSupportedTypes() {
       );
     }
+
+    shell.setHiveSessionValue(InputFormatConfig.SCHEMA_AUTO_CONVERSION, "true");
```
I'd probably make this a separate test case. No need to tack it on to the UnsupportedTypes test.

pvary replied:
Moved to a separate test case.

rdblue commented:
Overall, looks good to me. I raised a few minor points and there is a conflict in the serde.

pvary commented:
> Overall, looks good to me. I raised a few minor points and there is a conflict in the serde.

Thanks for the review.
Rebased and fixed to comments. (Hoping that I did not mess up something in the meantime 😄)

rdblue commented:
Thanks, @pvary!

pvary commented:
Thanks for the review and the merge, @rdblue!
